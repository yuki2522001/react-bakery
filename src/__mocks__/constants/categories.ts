import { CategoryProps } from "@common-types/category";

const CATEGORY_MOCKING_LIST: CategoryProps[] = [
  {
    id: "1",
    name: "Cake"
  },
  {
    id: "2",
    name: "Breads"
  }
]

export { CATEGORY_MOCKING_LIST };
