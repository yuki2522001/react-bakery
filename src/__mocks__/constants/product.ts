import { Product } from "@common-types/product";
import url from "../../assets/images/products/black-forest-cake.jpg";

const PRODUCT_MOCKING: Product = {
  id: "1",
  categoryId: "1651924356334",
  name: "Cheese cake",
  price: 20,
  quantity: 5,
  description: "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.",
  images: [
    url
  ]
}

const PRODUCT_MOCKING_LIST: Product[] = [
  {
    id: "1",
    categoryId: "1651924356334",
    name: "Cheese cake",
    price: 20,
    quantity: 5,
    description: "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.",
    images: [
      "/src/assets/images/products/cheese-cake.jpg",
      "/src/assets/images/products/cheese-cake.jpg",
      "/src/assets/images/products/cheese-cake.jpg"
    ]
  },
  {
    id: "2",
    categoryId: "1651924356334",
    name: "Amaretti Cookie",
    price: 10,
    quantity: 5,
    description: "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.",
    images: [
      "/src/assets/images/products/cheese-cake.jpg",
      "/src/assets/images/products/cheese-cake.jpg",
      "/src/assets/images/products/cheese-cake.jpg"
    ]
  }
]

export { PRODUCT_MOCKING, PRODUCT_MOCKING_LIST };
