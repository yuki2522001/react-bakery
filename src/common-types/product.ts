import { CategoryProps } from "./category";

export interface Product {
  id: string;
  name: string;
  price: number;
  images: string[];
  quantity: number;
  categoryId: string;
  description: string;
  product?: {};
}

export interface ProductContext {
  products: Product[];
  dispatch: Function,
  searchValue: string;
  setSearchValue: Function;
  categories: CategoryProps[];
  setCategories: Function;
}