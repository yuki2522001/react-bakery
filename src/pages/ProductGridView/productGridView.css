.product__list {
  padding: 0 100px;
}

.product__info {
  display: flex;
  flex-wrap: wrap;
}