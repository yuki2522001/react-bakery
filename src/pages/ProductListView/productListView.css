.viewProduct__list {
  display: flex;
  flex-wrap: wrap;
  padding: 30px 70px 40px 130px;
}

.viewProduct__link {
  text-decoration: none;
}