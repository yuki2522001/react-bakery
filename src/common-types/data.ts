import { Product } from "./product";

export enum Action {
  GetProductSuccess = "GET_PRODUCT_SUCCESS",
  CreateProductsSuccess = "CREATE_PRODUCT_SUCCESS",
  DeleteProductSuccess = "DELETE_PRODUCT_SUCCESS",
  UpdateProductSuccess = "UPDATE_PRODUCT_SUCCESS",
}

export interface DataState {
  products: Product[];
}

export type DataPayload = Product[] | string | Product

export interface DataAction {
  action: Action;
  payload: DataPayload
}
