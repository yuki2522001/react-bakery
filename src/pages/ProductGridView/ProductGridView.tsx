import useSWR from "swr";
import React, { useCallback, useContext, useEffect, useState } from "react";
import toast from "react-hot-toast";
import ModalCreate from "@components/Modal/ModalCreate/ModalCreate";
import ScrollButton from "@components/common/Button/ScrollButton/ScrollButton";
import ProductGridCard from "@components/ProductGridCard/ProductGridCard";
import { Product } from "@common-types/product";
import { SUCCESS_MSG } from "@constants/message";
import { create, getData, remove } from "@helpers/apiHandle";
import { PRODUCTS_URL } from "@constants/url";
import { DataContext } from "@context/DataContext";
import { Action } from "@common-types/data";
import "./productGridView.css";

const ProductGridView: React.FC = () => {
  const [openModalCreate, setOpenModalCreate] = useState<boolean>(false);
  const { products, searchValue, dispatch } = useContext(DataContext);
  const queryParams: URLSearchParams = new URLSearchParams(searchValue);
  const { data } = useSWR(PRODUCTS_URL + "?" + queryParams.toString(), getData<Product[]>);
  useEffect(() => {
    if(data) {
      dispatch({
        action: Action.GetProductSuccess,
        payload: data
      });
    }
  }, [data]);
  // create product
  const createProduct = async (productData: Product) => {
    const newProduct: Product = {
      id: new Date().valueOf().toString(),
      categoryId: productData.categoryId,
      name: productData.name,
      price: +productData.price,
      quantity: +productData.quantity,
      description: productData.description,
      images: productData.images,
    };
    try {
      const response = await create(PRODUCTS_URL, newProduct);
      if(response) {
        dispatch({
          action: Action.CreateProductsSuccess,
          payload: {...response.data}
        });
        toast.success(SUCCESS_MSG.MESSAGE_ADD_PRODUCT);
      }
    } catch (error: any) {
      toast.error(error.message);
    }
  };

  // delete product
  const deleteProduct = async (id: string) => {
    try {
      const response = await remove(`${PRODUCTS_URL}/${id}`);
      if(response) {
        dispatch({
          action: Action.DeleteProductSuccess,
          payload: id
        });
        toast.success(SUCCESS_MSG.MESSAGE_DELETE_PRODUCT);
      }
    } catch (error: any) {
      toast.error(error.message);
    }
  };

  // handle toggle modal create
  const toggleModalCreate = useCallback(() => {
    setOpenModalCreate(!openModalCreate);
  }, [openModalCreate]);

  return (
    <>
      <div data-testid="product-gird-view" className="product__list">
        <button
          data-testid="open-modal"
          className="btn btn__add"
          onClick={toggleModalCreate}
        >
          Add new product
        </button>
        <div data-testid="delete-product" className="product__info">
          {products?.map((product: Product) => (
            <div className="product__item" key={product.id}>
              <ProductGridCard
                product={product}
                deleteProduct={deleteProduct}
              />
            </div>
          ))}
        </div>
        {openModalCreate && (
          <ModalCreate
            createProduct={createProduct}
            hideModalCreate={toggleModalCreate}
          />
        )}
      </div>
      <ScrollButton className="btn__backToTop" children={<i className="fa fa-arrow-alt-circle-up"></i>} />
    </>
  );
};

export default ProductGridView;
