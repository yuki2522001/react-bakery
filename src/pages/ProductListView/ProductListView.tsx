import useSWR from "swr";
import React, { useContext, useEffect } from "react";
import Button from "@components/common/Button/Button/Button";
import ScrollButton from "@components/common/Button/ScrollButton/ScrollButton";
import ProductListCard from "@components/ProductListCard/ProductListCard";
import { Link } from "react-router-dom";
import { Product } from "@common-types/product";
import { DataContext } from "@context/DataContext";
import { Action } from "@common-types/data";
import { PRODUCTS_URL } from "@constants/url";
import { getData } from "@helpers/apiHandle";
import "./productListView.css";

const ProductListView: React.FC = () => {
  // URLSearchParams: convert searchValue to string => handle search
  const { products, searchValue, dispatch } = useContext(DataContext);
  const queryParams: URLSearchParams = new URLSearchParams(searchValue);
  const { data } = useSWR(PRODUCTS_URL + "?" + queryParams.toString(), getData<Product[]>);

  useEffect(() => {
    if(data) {
      dispatch({
        action: Action.GetProductSuccess,
        payload: data
      });
    }
  }, [data]);

  return (
    <>
      <div data-testid="view-product-list" className="viewProduct__list">
        {products?.map((product: Product) => (
          <div className="viewProduct__item" key={product.id}>
            <ProductListCard product={product} />
          </div>
        ))}
      </div>
      <Link className="viewProduct__link" to="/products">
        <Button type="secondary" text="VIEW ALL PRODUCTS" />
      </Link>
      <ScrollButton className="btn__backToTop" children={<i className="fa fa-arrow-alt-circle-up"></i>} />
    </>
  );
};

export default ProductListView;
