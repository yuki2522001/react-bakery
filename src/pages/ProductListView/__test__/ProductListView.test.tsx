import "@testing-library/jest-dom";
import mockAxios from "@__mocks__/axios";
import ProductListView from "../ProductListView";
import Categories from "@components/Categories/Categories";
import Button from "@components/common/Button/Button/Button";
import ProductGridView from "@pages/ProductGridView/ProductGridView";
import ModalUpdate from "@components/Modal/ModalUpdate/ModalUpdate";
import ModalCreate from "@components/Modal/ModalCreate/ModalCreate";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { Link, Router } from "react-router-dom";
import { CATEGORIES_URL, PRODUCTS_URL } from "@constants/url";
import { getData } from "@helpers/apiHandle";
import { CATEGORY_MOCKING_LIST } from "@__mocks__/constants/categories";
import { PRODUCT_MOCKING, PRODUCT_MOCKING_LIST } from "@__mocks__/constants/product";
import { ProductContext } from "@common-types/product";
import { DataContext } from "@context/DataContext";
import { Action, DataState } from "@common-types/data";
import { productReducer } from "@reducer/productReducer";

const contextProductMock: ProductContext = {
  products: PRODUCT_MOCKING_LIST,
  dispatch: jest.fn(),
  searchValue: "",
  setSearchValue: jest.fn(),
  categories: CATEGORY_MOCKING_LIST,
  setCategories: jest.fn()
};

describe("Product list view component", () => {
  afterEach(() => {
    mockAxios.reset();
  });

  test("get product list should call", async () => {
    mockAxios.get.mockResolvedValueOnce({ data: PRODUCT_MOCKING_LIST });
    const result = await getData(PRODUCTS_URL);
    expect(mockAxios.get).toHaveBeenCalledWith(PRODUCTS_URL);
    expect(result).toEqual(PRODUCT_MOCKING_LIST);
  });

  test("get categories should call", async () => {
    mockAxios.get.mockResolvedValueOnce({ data: CATEGORY_MOCKING_LIST });
    const result = await getData(CATEGORIES_URL);
    expect(mockAxios.get).toHaveBeenCalledWith(CATEGORIES_URL);
    expect(result).toEqual(CATEGORY_MOCKING_LIST);
  });

  test("should get product when run app", () => {
    mockAxios.get.mockResolvedValue({ data: PRODUCT_MOCKING_LIST });
    const history = createMemoryHistory();
    render(
      <DataContext.Provider value={contextProductMock}>
        <Router location={history.location} navigator={history}>
          <ProductListView />
        </Router>
      </DataContext.Provider>
    );
    waitFor(() => { expect(contextProductMock.dispatch).toHaveBeenCalled() });
  });

  test("should update product when click button submit", () => {
    mockAxios.put.mockResolvedValue(PRODUCT_MOCKING);
    render(
      <DataContext.Provider value={contextProductMock}>
        <ModalUpdate
          product={PRODUCT_MOCKING}
          hideModalUpdate={() => {}}
          deleteImage={() => {}}
          updateProductDetail={() => {}}
        />
      </DataContext.Provider>
    );
    const btnSubmit = screen.getByText("Submit");
    fireEvent.click(btnSubmit);
    waitFor(() => { expect(contextProductMock.dispatch).toHaveBeenCalled() });
  });

  test("should create product when click Submit", () => {
    mockAxios.post.mockResolvedValue(PRODUCT_MOCKING);
    render(
      <ModalCreate
        hideModalCreate={() => {}}
        createProduct={() => {}}
      />
    );
    const submitBtn = screen.getByText("Submit");
    fireEvent.click(submitBtn);
    waitFor(() => { expect(contextProductMock.dispatch).toHaveBeenCalled() });
  });

  test("should get product when run app", () => {
    mockAxios.get.mockResolvedValue({ data: PRODUCT_MOCKING_LIST });
    const history = createMemoryHistory();
    render(
      <DataContext.Provider value={contextProductMock}>
        <Router location={history.location} navigator={history}>
          <ProductGridView />
        </Router>
      </DataContext.Provider>
    );
    waitFor(() => {
      expect(contextProductMock.dispatch).toHaveBeenCalled();
    });
  });

  test("should get data when dispatch action GetProductSuccess", () => {
    const initialState: DataState = {
      products: PRODUCT_MOCKING_LIST,
    };
    const getProduct = {
      action: Action.GetProductSuccess,
      payload: PRODUCT_MOCKING_LIST,
    };
    const updatedState = productReducer(initialState, getProduct);
    expect(updatedState).toEqual(updatedState);
  });

  test("should filter when click category", () => {
    const { getByTestId } = render(
      <DataContext.Provider value={contextProductMock}>
        <Categories />
      </DataContext.Provider>
    );
    const categoryItem = getByTestId("category-item");
    fireEvent.click(categoryItem);
    expect(contextProductMock.setSearchValue).toHaveBeenCalled();
  });

  test("Redirect to product page when click button 'View all products'", async () => {
    const history = createMemoryHistory();
    history.push = jest.fn();
    const { getByText } = render(
      <Router location={history.location} navigator={history}>
        <Link to="/products">
          <Button type="secondary" text="VIEW ALL PRODUCTS" />
        </Link>
      </Router>
    );

    fireEvent.click(getByText("VIEW ALL PRODUCTS"));
    expect(history.push).toHaveBeenCalled();
  });

  test("matches snapshot", () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <DataContext.Provider value={contextProductMock}>
        <Router location={history.location} navigator={history}>
          <ProductListView />
        </Router>
      </DataContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
