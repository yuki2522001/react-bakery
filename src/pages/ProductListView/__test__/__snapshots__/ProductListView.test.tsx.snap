// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product list view component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="viewProduct__list"
    data-testid="view-product-list"
  >
    <div
      class="viewProduct__item"
    >
      <div
        class="productViewPage"
        data-testid="product-list-card"
      >
        <img
          class="productViewPage__image"
          src="/src/assets/images/products/cheese-cake.jpg"
        />
        <div
          class="productViewPage__content"
        >
          <a
            class="productViewPage__link"
            href="/products/1"
          >
            <h2
              class="productViewPage__title"
              data-testid="title-test"
            >
              Cheese cake
            </h2>
          </a>
          <p
            class="productViewPage__description"
            data-testid="text"
          >
            Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.
          </p>
          <p
            class="productViewPage__price"
            data-testid="price"
          >
            $20
          </p>
        </div>
      </div>
    </div>
    <div
      class="viewProduct__item"
    >
      <div
        class="productViewPage"
        data-testid="product-list-card"
      >
        <img
          class="productViewPage__image"
          src="/src/assets/images/products/cheese-cake.jpg"
        />
        <div
          class="productViewPage__content"
        >
          <a
            class="productViewPage__link"
            href="/products/2"
          >
            <h2
              class="productViewPage__title"
              data-testid="title-test"
            >
              Amaretti Cookie
            </h2>
          </a>
          <p
            class="productViewPage__description"
            data-testid="text"
          >
            Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.
          </p>
          <p
            class="productViewPage__price"
            data-testid="price"
          >
            $10
          </p>
        </div>
      </div>
    </div>
  </div>
  <a
    class="viewProduct__link"
    href="/products"
  >
    <button
      class="btn btn__secondary"
    >
      VIEW ALL PRODUCTS
    </button>
  </a>
  <button
    class="btn__backToTop"
    data-testid="back-to-top"
  >
    <i
      class="fa fa-arrow-alt-circle-up"
    />
  </button>
  <div />
</DocumentFragment>
`;
